connection/cluster/ClusterServerStub.java

```diff
@@ -21,36 +21,58 @@
 
 package com.vaticle.typedb.client.connection.cluster;
 
+import com.vaticle.typedb.client.api.connection.TypeDBCredential;
+import com.vaticle.typedb.client.common.exception.ErrorMessage;
+import com.vaticle.typedb.client.common.exception.TypeDBClientException;
 import com.vaticle.typedb.client.common.rpc.TypeDBStub;
 import com.vaticle.typedb.protocol.ClusterDatabaseProto;
 import com.vaticle.typedb.protocol.ClusterServerProto;
 import com.vaticle.typedb.protocol.ClusterUserProto;
+import com.vaticle.typedb.protocol.ClusterUserTokenProto;
 import com.vaticle.typedb.protocol.TypeDBClusterGrpc;
 import com.vaticle.typedb.protocol.TypeDBGrpc;
 import io.grpc.CallCredentials;
 import io.grpc.ManagedChannel;
 import io.grpc.Metadata;
+import io.grpc.Status;
+import io.grpc.StatusRuntimeException;
 
 import java.util.concurrent.Executor;
+import java.util.function.Supplier;
 
+import static com.vaticle.typedb.client.common.rpc.RequestBuilder.Cluster.ClusterUserToken.renewReq;
 import static io.grpc.Metadata.ASCII_STRING_MARSHALLER;
 
 public class ClusterServerStub extends TypeDBStub {
 
+    private final TypeDBCredential credential;
+    private String token;
+    private final ManagedChannel channel;
+    private final TypeDBGrpc.TypeDBBlockingStub blockingStub;
+    private final TypeDBGrpc.TypeDBStub asyncStub;
     private final TypeDBClusterGrpc.TypeDBClusterBlockingStub clusterBlockingStub;
 
-    private ClusterServerStub(ManagedChannel channel, TypeDBGrpc.TypeDBBlockingStub blockingStub, TypeDBGrpc.TypeDBStub asyncStub,
-                              TypeDBClusterGrpc.TypeDBClusterBlockingStub clusterBlockingStub) {
-        super(channel, blockingStub, asyncStub);
-        this.clusterBlockingStub = clusterBlockingStub;
+    ClusterServerStub(ManagedChannel channel, TypeDBCredential credential) {
+        super();
+        this.credential = credential;
+        this.channel = channel;
+        CredentialEmbedder credentialEmbedder = new CredentialEmbedder();
+        this.blockingStub = TypeDBGrpc.newBlockingStub(channel).withCallCredentials(credentialEmbedder);
+        this.asyncStub = TypeDBGrpc.newStub(channel).withCallCredentials(credentialEmbedder);
+        this.clusterBlockingStub = TypeDBClusterGrpc.newBlockingStub(channel).withCallCredentials(credentialEmbedder);
+        try {
+            ClusterUserTokenProto.ClusterUserToken.Renew.Res res = clusterBlockingStub.userTokenRenew(renewReq(this.credential.username()));
+            token = res.getToken();
+        } catch (StatusRuntimeException e) {
+            // ignore UNAVAILABLE exception
```

Comment:
Why would we want to ignore this exception? Surely a gRPC error will cause problems regardless of whether it is a connection error or not?

Reply:
What the stub is doing here is trying to **pre-acquire** a token of the user to a server such that by the time the user wants to do something, the token is already available.

If the request fail with `UNAVAILABLE`, it indicates a network failure or server crash, and as a result the token from that particular server will not be pre-acquired. We can ignore it since failure is a fact of life in TypeDB Cluster. Failing to pre-acquire a token just means the client will have to reattempt it later.